apiclient: add implicit default behavior for ephemeral-storage bind

ginglis13 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #630

**Description of changes:**

Add implicit default behavior where running `apiclient ephemeral-storage bind` without `--dirs` automatically uses platform-specific default directories.

Update help documentation to reflect the implicit default behavior where running `ephemeral-storage bind` without `--dirs` uses platform defaults.

**Testing done:**

Unit tests; Manual testing:

User data:

```
[settings.bootstrap-commands.ephemeral-storage]
commands = [
    ["apiclient", "ephemeral-storage", "init"],
    ["apiclient", "ephemeral-storage" ,"bind"]
]
essential = true
mode = "always"
```

k8s-1.33 variant `lsblk` output:

```
...
nvme3n1      259:16   0  6.8T  0 disk
`-md127        9:127  0 54.6T  0 raid0 /var/lib/kubelet
                                       /var/lib/soci-snapshotter
                                       /var/log/pods
                                       /var/lib/containerd
                                       /var/lib/host-containerd
                                       /mnt/.ephemeral

...
```


ecs-2 variant `lsblk` output:

```
...
nvme4n1      259:23   0  6.8T  0 disk
`-md127        9:127  0 54.6T  0 raid0 /var/lib/docker
                                       /var/log/ecs
                                       /var/lib/host-containerd
                                       /var/lib/containerd
                                       /mnt/.ephemeral
...
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 commented on sources/api/apiserver/src/server/ephemeral_storage.rs:
```diff
@@ -431,3 +442,43 @@ pub mod error {
+            .collect();
+
+        // Verify that k8s variant includes kubelet directory
+        assert!(expected_dirs.contains(&"/var/lib/kubelet".to_string()));
+        assert!(expected_dirs.contains(&"/var/lib/containerd".to_string()));
+        assert!(expected_dirs.contains(&"/var/lib/soci-snapshotter".to_string()));
+        assert!(expected_dirs.contains(&"/var/log/pods".to_string()));
```
Why the `to_string()` and then pass the reference of the `String`? Why not just `"/string"`?

ginglis13 replied:
`expected_dirs` is a `Vec<String>`. The `contains` method expects a reference to the container's type, so in this case, we must use `&String`, not `&str`, which are technically two different types

KCSesh replied:
You should be able to do something like - to clean it up a bit: 

```
for dir in [
        "/var/lib/kubelet",
        "/var/lib/containerd",
        "/var/lib/soci-snapshotter",
        "/var/log/pods"
] {
    assert!(allowed_dirs.allowed_exact.contains(dir));
}
```

Less String wrangling

ginglis13 replied:
@KCSesh this has the same `&str` vs `&String` issue

ginglis13 replied:
Ah I mistook that you've dropped `expected_dirs` and are running this check directly on `allowed_exact` - yup this works :D

ytsssun commented on sources/api/apiclient/src/main.rs:
```diff
@@ -794,9 +795,11 @@ fn parse_ephemeral_storage_init_args(args: Vec<String>) -> EphemeralStorageSubco
+    // If no arguments, use default directories
     if args.is_empty() {
-        usage_msg("Did not give arguments to bind")
+        return EphemeralStorageSubcommand::Bind(EphemeralStorageBindArgs { targets: None });
```
Can we just keep `targets` Vec<String> and just return empty vec here instead of wrapping it with the extra Option?

ytsssun commented on sources/api/apiclient/src/main.rs:
```diff
@@ -272,8 +272,9 @@ fn usage() -> ! {
-                                       (e.g. /var/lib/containerd). If no ephemeral disks are found
-                                       this operation does nothing.
+                                       (e.g. /var/lib/containerd). If not specified, uses platform
+                                       defaults. If no ephemeral disks are found this operation does nothing.
```
It seems that the "platform defaults" means more for ECS/EKS variants? Could we be more specific here?

If you don't think this is the best place, we might need to call this out somewhere else in the README.

ytsssun replied:
Not blocking: I also realize that we owe an update to the readme https://github.com/bottlerocket-os/bottlerocket-core-kit/blob/a54c56e11dd357e846a2450d89fdbac6401122a0/sources/api/apiclient/src/lib.rs#L1-L3 to include the ephemeral-storage subcommand there?

ginglis13 replied:
updated the helptext and added a commit w/ README changes for `ephemeral-storage`

ytsssun commented on sources/api/apiclient/src/main.rs:
```diff
@@ -1064,7 +1074,8 @@ async fn run() -> Result<()> {
             }
             EphemeralStorageSubcommand::Bind(bind_args) => {
-                ephemeral_storage::bind(&args.socket_path, bind_args.targets)
+                let targets = bind_args.targets.unwrap_or_default();
```
Looks like we want empty Vec anyway? To my question above.

ginglis13 replied:
Good callout, latest push takes this suggestion

ginglis13 committed:
apiclient: add implicit default behavior for ephemeral-storage bind

Add implicit default behavior where running 'apiclient ephemeral-storage
bind' without --dirs automatically uses platform-specific default
directories.

Update help documentation to reflect the implicit default behavior where
running 'ephemeral-storage bind' without --dirs uses platform defaults.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 committed:
apiclient: document ephemeral-storage commands in README

Add documentation for ephemeral-storage subcommands init, bind,
list-disks, and list-dirs.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 commented:
^ force push moves the README changes over to the `.tpl` and includes generated README

ytsssun commented on sources/api/apiclient/src/main.rs:
```diff
@@ -1064,7 +1070,8 @@ async fn run() -> Result<()> {
             }
             EphemeralStorageSubcommand::Bind(bind_args) => {
-                ephemeral_storage::bind(&args.socket_path, bind_args.targets)
+                let targets = bind_args.targets;
+                ephemeral_storage::bind(&args.socket_path, targets)
```
Nit: can drop the local variable `target` here? Essentially keep the original code.

ginglis13 commented:
^ force push takes @KCSesh suggestion for simpler unit test syntax

ytsssun commented:
> ^ force push takes @KCSesh suggestion for simpler unit test syntax

Might be worth mentioning is that one trick I learnt from @cbgbt is that, we could use the [#test_case](https://docs.rs/test-case/latest/test_case/) macro for tests like this where you have different test inputs that runs the same function.
